### ISSUE TITLE
"needs_db_update" may return incorrect flag

### ISSUE THREAD

**Issue by malayladu** [woocommerce/woocommerce#23089]
**Describe the bug**
I am just looking at the code of upgrade script of WooCommerce and in that I found that following function is being used in woocommerce/includes/class-wc-install.php file.

     private static function needs_db_update() {

        $current_db_version = get_option( 'woocommerce_db_version', null );

        $updates = self::get_db_update_callbacks();

        return ! is_null( $current_db_version ) && version_compare( $current_db_version, max( array_keys( $updates ) ), '<' );
    }   

In the above function it tries to find out available max updates using following code.

`max( array_keys( $updates ) `

So, in future if WooCommerce release 3.5.10 and then when above code will run then it will return 3.5.9 as max version even if 3.5.10 is available. 

So,  needs_db_update function will assume there is no update available even if 3.5.10 is available 

**To Reproduce**
1. Go to 'class-wc-install.php' file and add '3.5.10' in $db_updates variable and try to check whether it updates to 3.5.10 or not.

**Expected behavior**
If 3.5.10 is the latest version, this function should return true. Which means update is available.   

**WordPress Environment**
WordPress - 5.1.1
WooCommerce - 3.5.4

**Comment by mikejolley** [comment]
This is the `woocommerce_db_version` option. There is also a `woocommerce_version` option which tracks the actual version number. In this case, it's working as intended because the `max( array_keys( $updates )` part checks against the list of DB updates WC needs to perform.  If there are no newer updates (e.g. no db updates need to happen in 3.5.10) it is correct to return false.

tl;dr `needs_db_update` is only true if there are update callbacks to run for a specific version. cc @kloon

**Comment by malayladu** [comment]
@mikejolley  Yes..I know `needs_db_update ` function is for db updates.

What if 3.5.10 requires some db update? 

`max(array('3.5.4', '3.5.10'));` will return 3.5.4 instead of 3.5.10.

It's not a bug with a current version but it may occur as a bug if similar version requires to compare with `max` PHP function

**Comment by kloon** [comment]
Mike, this is definitely a bug which could cause DB updates to not run when we release any point fix that has a .10 and up. I tested it now by simulating a 5.4.10 db update and it gave false results, I have a patch ready which fixes this.

**Comment by mikejolley** [comment]
Thanks for clarifying @malayladu

**Comment by kloon** [comment]
See #23092

**Comment by malayladu** [comment]
@mikejolley 👍

**Comment by malayladu** [comment]
@kloon Thanks for the patch 👍
